plugins/module_utils/ldap.py

```diff
@@ -60,6 +60,8 @@ def _connect_to_ldap(self):
             ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
 
         connection = ldap.initialize(self.server_uri)
+        # Switch off chasing of referrals (https://github.com/ansible-collections/community.general/issues/1067)
+        connection.set_option(ldap.OPT_REFERRALS, 0)
```

Comment:
Right now, you are setting this option for **all** LDAP modules. Is this intentional? Could this break other modules?

Reply:
No, it is not intentional. You are right, other modules (like ldap_attr) are using search_s ldap library function. To avoid potential modules break i suggest to add an optional module parameter named referrals_chase. If set, this parameter switch off referrals chasing. If not, the default behavior is kept, as now.

Reply:
Why not simply have a property `disable_referrals` with default value `False`, which is set to `True` by the `ldap_search` module? That way, you can disable it for the `ldap_search` module without disabling it for the other modules, and without introducing a new user-facing parameter.

Reply:
The [python-ldap](https://www.python-ldap.org/en/python-ldap-3.3.0/faq.html) documentation explains that by default when an LDAP server (A) returns a referral to another LDAP server (B), the client sends the request anonymously to server B. If server B accepts requests anonymous then server B returns the result of the request to the client, otherwise the request is in error. In the case of an AD server the request will be systematically in error because AD generally does not allow anonymous authentication.
By disabling referral chasing at the module level, the risk is to break the operation of a current playbook that would track referrals to LDAP servers accepting anonymous requests. That's why I suggested using an optional attribute to turn off referral chasing.

Reply:
@lamm you could add an optional parameter for the `ldap_search` module, but you still have to make sure that the other modules are not affected by this change.

Reply:
@felixfontein If the new parameter (referrals_chasing) is omitted, nothing is changed to the current modules behaviour because we set _OPT_REFERRALS_ to 0 only if **referrals_chasing** is set to _'disabled'_.
Modifying module_utils/ldap.py allows to use referrals_chasing parameter to modify the default behaviour for the other ldap modules in case of the same search issue is encountered.
Instead of using a boolean for referrals_chasing parameter type, i implemented a list of choice. This allow to implement a new behaviour in the future (like following referrals with different credentials).
I also modified module/ldap_search.py to fix the error encountered when ldap_search module try to parse the referral as an ldap entry #963.
Do you agree with that ? Let me know if modifications needed.